### PR TITLE
 Implementation of periodic boundary condition for DG advection solver

### PR DESCRIPTION
**Pull request by yinghe616**
Initial implementation of periodic boundary condition for the DG advection solver

### CODE REVIEW

File: source/simulator/assemblers/advection.cc [hunk 1]
```diff
@@ -479,23 +484,43 @@ namespace aspect
...
-          if (!cell->neighbor_is_coarser(face_no) &&
+          const bool neighbor_or_periodic_neighbor_is_coarser =
+            ( cell_has_periodic_neighbor
+              ?
+              cell->periodic_neighbor_is_coarser(face_no)
+              :
+              cell->neighbor_is_coarser(face_no));
+
+          if (!neighbor_or_periodic_neighbor_is_coarser &&
+              // neighbor->level () == cell->level () &&
```
Comment (yinghe616) [inline source/simulator/assemblers/advection.cc]: I defined a similar bool constant "neighbor_or_periodic_neighbor_is_coarser" as the original "neighbor_is_coarser", but I just realized in the line 506, calling neighbor->level () == cell->level () does the same as calling " !neighbor_or_periodic_neighbor_is_coarser "  (are they the same?), but the former call doesn't need to define an extra bool constant "neighbor_or_periodic_neighbor_is_coarser"

So I personally prefer to use only neighbor->level () == cell->level (),  any comment?

File: source/simulator/assemblers/advection.cc [hunk 2]
```diff
@@ -437,6 +437,11 @@ namespace aspect
                 }
             }
         }
+      else if (cell->has_periodic_neighbor (face_no))
```
Comment (bangerth) [inline source/simulator/assemblers/advection.cc]: will this lead to assembling the term twice, once from each side of the periodic boundary?

File: source/simulator/assemblers/advection.cc (hunk 1 above)
Comment (bangerth) [inline source/simulator/assemblers/advection.cc]: your choice -- pick one, but don't leave a commented out line

File: source/simulator/assemblers/advection.cc [hunk 3]
```diff
@@ -789,14 +814,24 @@ namespace aspect
...
+          const unsigned int neighbor2 =
+            (cell_has_periodic_neighbor
+             ?
+             cell->periodic_neighbor_face_no(face_no)
+             :
+             cell->neighbor_face_no(face_no));
 
           //loop over subfaces
+          std::cout<< "cell no " << cell  << std::endl;
+          std::cout<< "yyyy test" << std::endl;
```
Comment (bangerth) [inline source/simulator/assemblers/advection.cc]: Please remove.

File: source/simulator/assemblers/advection.cc (hunk 2 above)
Comment (yinghe616) [inline source/simulator/assemblers/advection.cc]: I think we had agreed with the changes. I just added a simple test.

File: source/simulator/assemblers/advection.cc (hunk 1 above)
Comment (yinghe616) [inline source/simulator/assemblers/advection.cc]: addressed

File: tests/prescribed_stokes_solution_DG_periodic.prm [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+end
+
+
+subsection Initial temperature model
+  set Model name = function
+
+  subsection Function
+    set Variable names      = x,y
+    set Function constants  = pi=3.1415926536
+    set Function expression = cos(2*pi*x)*sin(2*pi*y)
```
Comment (bangerth) [inline tests/prescribed_stokes_solution_DG_periodic.prm]: yes, nice

**Comment by bangerth** [comment]
This looks good. OK to merge once the tester is happy.

**Comment by yinghe616** [comment]
The tester returned this error message, but I think I only added a test example with prescribed Stokes solution... so I couldn't figure out what reason cause this error.
http://www.math.clemson.edu/~heister/cib/aspect-8.5-gcc-hd/ab382a08ac54131076e7cf0be6da7bba6d2acd5d/
BUILD gcc: 
FAILED: output-n_expensive_stokes_solver_steps/solver_history.txt.diff 
test FAILED 
ok 
tests: 1 / 435 FAILED 
DIFFS: changes-gcc.diff

**Comment by gassmoeller** [comment]
Ying could you update your master branch and rebase this pull request to that one? The test failure is likely unrelated to your pull request, and should be solved in the recent master.

**Comment by yinghe616** [comment]
Done rebasing.

**Pull request merged**